integration/client.go

```diff
@@ -0,0 +1,128 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package integration
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3metrics/protocol/msgpack"
+)
+
+type client struct {
+	address        string
+	batchSize      int
+	connectTimeout time.Duration
+	encoder        msgpack.UnaggregatedEncoder
+	conn           net.Conn
+}
+
+func newClient(address string, batchSize int, connectTimeout time.Duration) *client {
+	return &client{
+		address:   address,
+		batchSize: batchSize,
+		encoder:   msgpack.NewUnaggregatedEncoder(msgpack.NewPooledBufferedEncoder(nil)),
+	}
+}
+
+func (c *client) connect() error {
+	conn, err := net.DialTimeout("tcp", c.address, c.connectTimeout)
+	if err != nil {
+		return err
+	}
+	c.conn = conn
+	return nil
+}
+
+func (c *client) testConnection() bool {
+	if err := c.connect(); err != nil {
+		return false
+	}
+	c.conn.Close()
+	c.conn = nil
+	return true
+}
+
+func (c *client) write(mu unaggregated.MetricUnion, vp policy.VersionedPolicies) error {
+	encoder := c.encoder.Encoder()
+	sizeBefore := encoder.Buffer.Len()
+	var err error
+	switch mu.Type {
+	case unaggregated.CounterType:
+		err = c.encoder.EncodeCounterWithPolicies(unaggregated.CounterWithPolicies{
+			Counter:           mu.Counter(),
+			VersionedPolicies: vp,
+		})
+	case unaggregated.BatchTimerType:
+		err = c.encoder.EncodeBatchTimerWithPolicies(unaggregated.BatchTimerWithPolicies{
+			BatchTimer:        mu.BatchTimer(),
+			VersionedPolicies: vp,
+		})
+	case unaggregated.GaugeType:
+		err = c.encoder.EncodeGaugeWithPolicies(unaggregated.GaugeWithPolicies{
+			Gauge:             mu.Gauge(),
+			VersionedPolicies: vp,
+		})
+	default:
+		err = fmt.Errorf("unrecognized metric type %v", mu.Type)
+	}
+	if err != nil {
+		encoder.Buffer.Truncate(sizeBefore)
+		c.encoder.Reset(encoder)
+		return err
+	}
+	sizeAfter := encoder.Buffer.Len()
+	// If the buffer size is not big enough, do nothing
+	if sizeAfter < c.batchSize {
+		return nil
+	}
+	// Otherwise we get a new buffer and copy the bytes exceeding the max
+	// flush size to it, swap the new buffer with the old one, and flush out
+	// the old buffer
+	encoder2 := msgpack.NewPooledBufferedEncoder(nil)
+	data := encoder.Bytes()
+	encoder2.Buffer.Write(data[sizeBefore:sizeAfter])
+	c.encoder.Reset(encoder2)
+	encoder.Buffer.Truncate(sizeBefore)
+	_, err = c.conn.Write(encoder.Bytes())
```

Comment:
Can we set a write timeout that's passed in via options. We don't want to use default vals here for when server side dies in certain ways.